Proposal: Introduce Custom Metrics API

DirectXMan12 opened the pull request:
**Release note**:

``` release-note
NONE
```

This proposal details the custom metrics API as proposed in the new
monitoring vision.  It is designed for use with the HPA, but should be
generally useful for controllers that wish to consumer custom metrics.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34586)

<!-- Reviewable:end -->

DirectXMan12 committed:
Proposal: Introduce Custom Metrics API

This proposal details the custom metrics API as proposed in the new
monitoring vision.  It is designed for use with the HPA, but should be
generally useful for controllers that wish to consumer custom metrics.

DirectXMan12 commented:
cc @kubernetes/autoscaling 

This is my initial proposal for roughly what the new HPA custom metrics API proposed in the monitoring vision would look like.  I've tried to capture the potential use cases as well as some of the concerns around the actual machinery.  Let me know if I missed any.

DirectXMan12 commented on docs/proposals/custom-metrics-api.md:
```diff
@@ -0,0 +1,189 @@
+- `/namespaces/{namespace-name}/metrics/{metric-name...}`: retrieve the
+  given metric on the given namespace
+
+- `/namespaces/{namespace-name}/object-metrics/{object-type}/{metric-name...}`:
```
I should probably clarify what `object-type` means in a multi-API-group world -- I was thinking that it would be the string form of the `GroupKind` object in Kubernetes.  This does potentially introduce the requirement to recognize objects moved between API groups, but I don't think there's any good way around that.

balajismaniam commented:
cc @andrzej-k @IzabellaRaulin.

piosz commented:
cc @fabxc

fabxc reviewed (commented):
Thank you for writing this up and clarifying the whole picture.

Practically, the proposed API seems to allow to query a mapping of actual queries. The name "metric" here is rather misleading. Any non-naive monitoring system will have dimensionality below the pod level, aggregations over them, transforming functions etc. etc.
Obviously even more when talking about deployments, services, ... That means up to hundreds of thousands of time series per "metric". The API cannot seem to express that.
So an implementing gateway to my monitoring system would effectively have to store a mapping to an actual query it performs. The lookup key would then be what this API defines as "metric". 

I'd have to write a query for each monitoring system regardless – would a simple generic query-forwarder be simpler and feature equivalent? Binding it to resource objects seems somewhat artificial.
But I lack knowledge about how the HPA would consume them to judge that.

 👍 for not introducing a meta query language.

A more general thought:

I'm always wondering about how sophisticated an organisation has to be before having a practical need for the use cases mentioned for custom HPA metrics. Way before that, responsiveness of basic auto-scaling seem to become important. 
Not sure whether it was discussed yet, but were there concerns about the staleness of metrics by the time they reach the consumer? With all the propagation through various systems 2-5 minutes won't be uncommon. I'd consider whether monitoring systems (or anything async) are the best source for auto-scaling metrics in the first place.

Slightly off topic:

I'm concerned about the idea of extending the usage of such an API to `kubectl top` or the Dashboard UI. These are both monitoring tools.
There's no value in saying we are agnostic about the monitoring systems when we then define very limited interfaces all monitoring systems simply have to fit into. Just making an authoritative choice of a default monitoring system would be better – at least that could be used to its full capabilities.
I think the original reason to be agnostic was A) that monitoring is a hard problem that B) Kubernetes should not aim to solve because C) many projects worked long and hard to find solutions for said problems, many of them generally incompatible with each other.
If we want to be agnostic, why don't we let monitoring projects find good solutions for above monitoring problems that leverage their systems to the best possible degree? Being agnostic to me means having multiple dashboarding, `top`, and alerting solutions. And that's okay.

Looking at Heapster, (at least some) people also said it was not meant for monitoring. Today it effectively is the only (very limited) monitoring of their cluster to most users and it certainly inhibited advancements by eco-systems of dedicated monitoring projects. I don't want to see the same happening for custom metrics.

DirectXMan12 commented on docs/proposals/custom-metrics-api.md:
```diff
@@ -0,0 +1,189 @@
+    "metricValues": [
+        {
+            "kind": "Pod",
+            "metadata": { "name": "server1", "namespace": "webapp" },
```
It doesn't currently include it in the master metrics API -- we get to choose which bits an pieces we actually populate, although it isn't expressed at the type level.  It might be advantageous to express the relationship using a objectreference instead, but I was going for symmetry with the master/resource metrics API.  I'm not opposed to changing this, but I have a preference for keeping them relatively similar in shape.

DirectXMan12 commented on docs/proposals/custom-metrics-api.md:
```diff
@@ -0,0 +1,189 @@
+}
+
+// a metric value for some object
+type GenericMetrics struct {
```
whoops, there's a typo above -- it should list `GenericMetrics` instead of `MetricsValue` above (to match the name scheme from the master metrics API -- we have NodeMetrics, PodMetrics, and GenericMetrics).

DirectXMan12 commented:
> Any non-naive monitoring system will have dimensionality below the pod level,
> aggregations over them, transforming functions etc. etc. Obviously even more
> when talking about deployments, services, ... That means up to hundreds of
> thousands of time series per "metric". The API cannot seem to express that.
> So an implementing gateway to my monitoring system would effectively have to
> store a mapping to an actual query it performs. The lookup key would then be
> what this API defines as "metric".

Yeah, part of the idea here is that the API abstracts over the details of the
backend, so that Kubernetes-related entities can talk in terms of things that
they understand easily -- Kubernetes-related-things want to ask "what are the
hits-per-second recorded by the ingress controller for this Ingress object (or
service)" or "for this group of pods, what is the
queries-processed-per-second".

> I'd have to write a query for each monitoring system regardless – would
> a simple generic query-forwarder be simpler and feature equivalent? Binding
> it to resource objects seems somewhat artificial.
But I lack knowledge about how the HPA would consume them to judge that.

I'm not quite sure what you're suggesting here, but any system we have needs to
not require the HPA to know any specifics of the monitoring system, because we
don't want to have to recompile the controller manager to support different
metrics systems (and because that makes other controllers and tools consuming
this in the future harder).

> A more general thought:

> I'm always wondering about how sophisticated an organisation has to be before
> having a practical need for the use cases mentioned for custom HPA metrics.
> Way before that, responsiveness of basic auto-scaling seem to become
> important.

> Not sure whether it was discussed yet, but were there concerns about the
> staleness of metrics by the time they reach the consumer? With all the
> propagation through various systems 2-5 minutes won't be uncommon. I'd
> consider whether monitoring systems (or anything async) are the best source
> for auto-scaling metrics in the first place.

That's definitely a concern (and we have discussed metrics latency in the
past).  What kind of latency do you usually see on a common Prometheus setup?
I'd be curious to see some measurements from common systems.  Depending on
setup size, I think we've gotten pretty decent latency with
Kubelet->Heapster->Hawkular (assuming a sufficiently quick interval for
Heapster, like 15s). @mwringe can probably give better details there.

One of the nice things here is that since this will be a standard API, we have
the freedom to implement it on top off different systems.  If your metrics
system handles the latency you need, that's great, if you have tight latency
requirements and/or your metrics system isn't really designed for that, you
could use a separate store, etc.

> Slightly off topic:

> I'm concerned about the idea of extending the usage of such an API to kubectl
> top or the Dashboard UI. These are both monitoring tools.

I view the dashboard more as a UI that has some basic monitoring display involved
(see below).

> There's no value in saying we are agnostic about the monitoring systems when
> we then define very limited interfaces all monitoring systems simply have to
> fit into. Just making an authoritative choice of a default monitoring system
> would be better – at least that could be used to its full capabilities.

> I think the original reason to be agnostic was A) that monitoring is a hard
> problem that B) Kubernetes should not aim to solve because C) many projects
> worked long and hard to find solutions for said problems, many of them
> generally incompatible with each other.

> If we want to be agnostic, why don't we let monitoring projects find good
> solutions for above monitoring problems that leverage their systems to the
> best possible degree? Being agnostic to me means having multiple
> dashboarding, top, and alerting solutions. And that's okay.

I think it's important to give monitoring solution authors good tools and
integration points to hook into Kubernetes, because I definitely agree there
should be multiple options for monitoring, and we should make it "easy" for
people to build sophisticated monitoring solutions on top of Kubernetes.

> Looking at Heapster, (at least some) people also said it was not meant for
> monitoring. Today it effectively is the only (very limited) monitoring of
> their cluster to most users and it certainly inhibited advancements by
> eco-systems of dedicated monitoring projects. I don't want to see the same
> happening for custom metrics.

I think there are a couple reasons that Heapster became such a bottleneck for
development.  First, Heapster was required for HPA.  The new vision attempts to
get rid of that requirement.  Second, Heapster was billed as an easy-to-set-up
solution from monitoring, and since it already knew how to talk to kubelet and
put those metrics in a usable form.  I think having some "ready-made" solutions
besides heapster would go a long way.

On the other hand, I think it could be beneficial to have some APIs for
basic functionality, if you're writing something that's cluster-agnostic
or not *primarily* a monitoring tool. If you want to build a whole
monitoring dashboard which hooks into your monitoring solution and
displays customized graphs, lets you build fancy queries, etc go ahead. If
you're building a snazzy UI for Kubernetes, and want to display basic
metrics along side your pod listings, I see little reason not to have an
easy API to hook in to.

From a more practical perspective, a) we already have kubectl top, and b)
there's little we can do to prevent people from using these APIs.

fgrzadkowski commented on docs/proposals/custom-metrics-api.md:
```diff
@@ -0,0 +1,189 @@
+
+`/apis/general-metrics/v1alpha1/namespaces/webapp/object-metrics/pods/hits-per-second?labelSelector=app%3Dfrontend`.
+
+API Objects
```
Is this API supposed to return the most recent value of a given metric? How does it combine multiple timeseries? For example let's say we expose a metric with request errors with some additional labels:
* request path
* request size
* which backend we used

In HPA we want to scale if number of all 200s is above 57 per instance. How can I express this using this API?

DirectXMan12 replied:
> Is this API supposed to return the most recent value of a given metric?

Yes, just like the master metrics API.

> In HPA we want to scale if number of all 200s is above 57 per instance. How can I express this using this API?

> How does it combine multiple timeseries?

The HPA v2 has no notion of labels, so I did not put any notion of labels into this API.  My initial gut reaction is to say that the adapter should have a way to expose labeled metrics within this schema, but I'm open to ideas.

fgrzadkowski commented on docs/proposals/custom-metrics-api.md:
```diff
@@ -0,0 +1,189 @@
+It shares many mechanical requirements with normal Kubernetes APIs.
+
+The API should support encoding the result in both JSON and Protobuf,
+similarly to how core Kubernetes APIs work.  Most likely, the Protobuf
```
I don't think we need Protobuffs any time soon. I think that the latency in monitoring pipeline itself will be large enough that we will not care that much about performance of communication protocol. This could become a problem in huge clusters where everything is autoscaled, but we should start with something simpler IMO.

DirectXMan12 replied:
this was explicitly suggested by some of the SIG API team, IIRC.  It doesn't have to be there initially (since we have a method of indicating whether or not protobuf is supported), but I think we should plan on putting it in.

fgrzadkowski commented on docs/proposals/custom-metrics-api.md:
```diff
@@ -0,0 +1,189 @@
+### Discoverability ###
+
+While the initial plan for the API is for consumption by the HPA
+controller, the API could be more broadly useful -- custom autoscaler or
```
We should focus only on HPA use-case because this is the only one that we understand. Desiging it more generic will not result in anything good - it will not be good for HPA and most likely also not for other use-cases as we don't understand them. I'd suggest being explicit in this proposal. Obviously we should not disallow people to use it, but there's no reason in encourage them either.

DirectXMan12 replied:
I know at the moment we have users who have written applications against the Heapster model API, so I think we should keep in mind that others might want to use the API.

That being said, I agree that it shouldn't be too generic.  The only thing in this API that the HPA v2 can't use is selecting multiple objects that aren't pods, but preventing this just makes the API more complicated.

fgrzadkowski commented on docs/proposals/custom-metrics-api.md:
```diff
@@ -0,0 +1,189 @@
+scheduler implementations could make use of it's structured presentation
+of metrics.  Broadly speaking, any consumer that might wish to
+consume "core" metrics could potentially benefit from also being able to
+consume general metrics.
```
IMO it'd be extremely rare so let's not make this proposal too generic.

DirectXMan12 replied:
(see above)

fgrzadkowski commented on docs/proposals/custom-metrics-api.md:
```diff
@@ -0,0 +1,189 @@
+and summarization plans, as with the master metrics API).
+
+The API should thus be discoverable and accessible as if it were just
+another API server, similarly to the master metrics API.
```
IMO it should not. Instead I'd suggest starting small - create a service inside of a cluster that will implement a well define API. It'll be enough to just pass name of this service to HPA and that's it. Other clients can do the same thing, but I don't see any reason to make it more discoverable.

DirectXMan12 replied:
I'm fine with having the initial implementation work like this, but I feel that we should eventually move to using discovery, so I think it should be in proposal at some point.

fgrzadkowski commented on docs/proposals/custom-metrics-api.md:
```diff
@@ -0,0 +1,189 @@
+
+Additionally, the ability to query for metrics for multiple objects in
+a given namespace is useful for dashboard-like interfaces (e.g.
+a dashboard UI, a top-like CLI, etc).
```
Please don't mention other use-cases as we are not targeting them.

fgrzadkowski reviewed (commented):
My main concern is that this API is not concrete enough. On one hand, we have a very concrete use-case in HPA that I think cannot be easily expressed using this API. On the hand someone who would have to implement this API would have a hard time understanding what is the semantics.

Just to be clear - I think we need to rethink Master Metric API as well as it's also not very good.

DirectXMan12 commented on docs/proposals/custom-metrics-api.md:
```diff
@@ -0,0 +1,189 @@
+Custom Metrics API
+==================
+
+The new [metrics monitoring vision]() proposes an API that the Horizontal
```
will do (it wasn't merged yet when I posted this).

DirectXMan12 commented:
> My main concern is that this API is not concrete enough. On one hand, we have a very concrete use-case in HPA that I think cannot be easily expressed using this API.

This statement is very confusing to me, as the API paths and arguments basically correspond one-to-one with the HPA v2 volume types:

- `{"type": "pods", "pods": {"metricName": "my-metric"}}` becomes `/apis/generic-metrics/v1alpha1/namespaces/$NS/object-metrics/pod/my-metric?labelSelector=$SCALE_TARGET_SELECTOR`
- `{"type": "object", "object": {"metricName: "my-metric", "target": {"kind": "Service", "name": "my-service"}}` becomes `/apis/generic-metrics/v1alpha1/namespaces/$NS/object-metrics/service/my-metric?names=my-service`
- namespace metrics (however we choose to represent them) become `/apis/generic-metrics/v1alpha1/namespaces/$NS/metrics/my-metric`

> On the hand someone who would have to implement this API would have a hard time understanding what is the semantics.

> Just to be clear - I think we need to rethink Master Metric API as well as it's also not very good.

Can you clarify what you mean here?  I have some thoughts on the master metrics API, but it generally seems mostly reasonable.

fabxc commented:
> I'm not quite sure what you're suggesting here, but any system we have needs to
not require the HPA to know any specifics of the monitoring system, because we
don't want to have to recompile the controller manager to support different
metrics systems (and because that makes other controllers and tools consuming
this in the future harder).

If a specific metric is always an end result of a more complex query over the actual metrics, the API seems to rather be about query indirection (via an identifier, called metric here).
For Prometheus that actually maps nicely to [recording rules](https://prometheus.io/docs/practices/rules/). So when implementing a Prometheus gateway, I'd just allow "metric" to be any Prometheus query which is directly forwarded and the result returned in an altered format.
Just that the result format proposed does not allow to propagate the label dimensionality and forces my gateway to restore mapping to k8s resource objects from the result data, which is hard. So I'm wondering whether the HPA actually needs the association or the information "result > x", where result can be about anything, is sufficient and actually more powerful/flexible.

> What kind of latency do you usually see on a common Prometheus setup?

It depends. Most important is the scrape interval, if it's 5s vs 60s obviously dictates the minimum upper bound. But then there are further factors. A monitoring system tolerates missing a scrape occasionally. Recording rules (mentioned above) may add up to one rule evaluation interval of delay (30-60s typically). Federation of said recording rules to the server the HPA actually gets its metrics from may add another one. The gateway may add another one.

Typically, you'd probably end up in the 15s range too. But that's not guaranteed. I wouldn't count on the operators of the Prometheus setup to always adhere to what the HPA (a single user of the system) wants.
So it's a question of which guarantees the HPA would like to have.

Agreed that it's flexible as nobody dictates to use data from a monitoring system.

> On the other hand, I think it could be beneficial to have some APIs for
basic functionality, if you're writing something that's cluster-agnostic
or not primarily a monitoring tool. If you want to build a whole
monitoring dashboard which hooks into your monitoring solution and
displays customized graphs, lets you build fancy queries, etc go ahead. If
you're building a snazzy UI for Kubernetes, and want to display basic
metrics along side your pod listings, I see little reason not to have an
easy API to hook in to.

I do agree. But I don't see custom metrics fitting into that. By definition, they are not generic and available in every cluster. So I don't see how they help for building cluster-agnostic dashboards for example.
The master metrics API provides the core metrics for that. Everything that's not core cannot be assumed to exist.

Same for `kubectl top`. I'm fine that it exists but I don't see how this can or should realistically be expanded beyond core metrics.

DirectXMan12 commented:
> If a specific metric is always an end result of a more complex query over the actual metrics, the API seems to rather be about query indirection (via an identifier, called metric here).

I think it depends on the monitoring system and it's storage schema, as well as the particular query (for instance, if we produce an aggregated view of pods up to RCs automatically, and we're in Hawkular, a metrics request for a metric on a single RC object becomes a single HTTP request).

> For Prometheus that actually maps nicely to recording rules. So when implementing a Prometheus gateway, I'd just allow "metric" to be any Prometheus query which is directly forwarded and the result returned in an altered format.

I'd be somewhat nervous allowing arbitrary queries through -- among other things, this makes it hard (or impossible) to filter on access controls (plus, it places a somewhat heavy burden on the user of the HPA).

> Just that the result format proposed does not allow to propagate the label dimensionality and forces my gateway to restore mapping to k8s resource objects from the result data, which is hard.

Can you not structure like this: I ask for metric "foo" on pods "bar", "baz", "quux" in namespace "ns" you submit 3 queries together like "select foo where namespace=ns and pod=bar; select foo where namespace=ns and pod=baz; select foo where namespace=ns and pod=quux" (apologies, I'm not familiar with the Prometheus query format, so I've structure these in a pseudo-SQL-style query)?  I agree having three separate queries is somewhat unsatisfying, but could be mitigated in certain situations by the consumer asking for the metric aggregated up to a single controller instead.

> So I'm wondering whether the HPA actually needs the association or the information "result > x", where result can be about anything, is sufficient and actually more powerful/flexible.

Well, "result > x" is not actually what we want, but we may be able to tweak the response format to make it easier to compose the queries:

What the HPA needs, essentially, is "what is the average of the values of this metric for *ready* pods matching the given label selector" and "what is the value of this metric for the given other object or objects".  The assumption in the current API is that the HPA controller will do the actual filtering on ready vs unready, as well as the averaging, and that the adapter API will simply return the relevant metrics by pod name.

If we look at the first query type ("what is the average..."), we could hypothetically, only return a single data point, the average.  This could pose a problem, depending on how we do it, though: it either means that the API needs to understand ready vs unready, or it must take in name lists only.

In the first case, the HPA sends a label selector, and the adapter converts this to a list of ready pods.  The problem I see there is that this ties the HPA logic to the adapter API, which in turn means that if we ever want to change how that works in the HPA, we need to also change the adapter API, which is not a desirable situation.

On the other hand, in the latter case, we need the HPA to convert a label selector to a pod list ahead of time (which isn't too bad, honestly).  It also means we need to make pods a separate case in the API from objects, and we lose the similarity to the other metrics APIs.  It's an interesting tradeoff, though.  WDYT?

> Typically, you'd probably end up in the 15s range too. But that's not guaranteed. I wouldn't count on the operators of the Prometheus setup to always adhere to what the HPA (a single user of the system) wants.
> So it's a question of which guarantees the HPA would like to have.

Yeah, I guess my opinion on that would be that it's basically up to the cluster admins to set up monitoring according to the level of service that they want to guarantee from the HPA controller, and make that known to the users ("we guarantee that custom metrics will be at least X fresh" or "expect to see this much time between metric changes and scale").

DirectXMan12 commented:
> The master metrics API provides the core metrics for that. Everything that's not core cannot be assumed to exist.

> Same for kubectl top. I'm fine that it exists but I don't see how this can or should realistically be expanded beyond core metrics.

it's fair point.  I'd perhaps point out that there's also a convenience factor here, though -- being able to ask for a metric by name and namespace or label selector is nice (as opposed to having to deal with the query language specifics of your monitoring system), but I think dashboards would be more likely to want that out of a historical API than this one.  I do think we should allow other uses of this API, but it's probably less likely to be dashboards.  I'll remove/tweak that language in the proposal when I do the next pass of revisions.

smarterclayton commented:
If we are adding apis to Kubernetes, they should follow the Kubernetes
conventions as much as possible and reuse common machinery.  I do not want
to worry about optimizing two stacks for apis to meet performance goals.
We'd need a really good reason *not* to have metrics APIs reusing core
machinery.

On Oct 26, 2016, at 11:26 AM, Solly Ross <notifications@github.com> wrote:

The master metrics API provides the core metrics for that. Everything
that's not core cannot be assumed to exist.

Same for kubectl top. I'm fine that it exists but I don't see how this can
or should realistically be expanded beyond core metrics.

it's fair point. I'd perhaps point out that there's also a convenience
factor here, though -- being able to ask for a metric by name and namespace
or label selector is nice (as opposed to having to deal with the query
language specifics of your monitoring system), but I think dashboards would
be more likely to want that out of a historical API than this one. I do
think we should allow other uses of this API, but it's probably less likely
to be dashboards. I'll remove/tweak that language in the proposal when I do
the next pass of revisions.

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/34586#issuecomment-256382068>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p9Z1y1jZtDiyR-PDOlzrY_Ta5zf6ks5q33EQgaJpZM4KUVNi>
.

DirectXMan12 commented:
I've updated the proposal a bit based on what we talked about in today's SIG autoscaling meeting.  The metric entries now hold `ObjectReferences` instead of type- and object-metadata, in order to more accurately represent our usage.  I've added a section clarifying the semantics, noted that the API design is primarily motivated by the HPA (and how the API maps to HPA source types), discussed how to deal with labeled metrics, and removed language about `kubectl top` and friends.  Please let me know if I missed anything.

DirectXMan12 commented:
@thockin those names look a bit better? I updated the names of the targets in the `pod` and `resource` types.  I've been trying to come up with a better name for the `resource` type, but most of the names seem verbose or awkward, and it's probably going to be the most common type, so it should be quick to read and type.  Currently, it will look like:

```yaml
metrics:
- type: resource
  resource:
    name: cpu
    targetAverageRequestUtilization: 80
`

DirectXMan12 commented:
@piosz this look good based on our discussions at Kubecon?

DirectXMan12 commented:
closing in favor of kubernetes/community#152
